R/mobr_boxplots.R

```diff
@@ -143,240 +170,270 @@ get_pval <- function(rand, obs, n_samples)
-      F_rand$beta_ENS_PIE[i]  = anova(lm(beta_ENS_PIE ~ group_id_rand))$F[1]
+   # Asymptotic estimates species richness -------------------------------------
+  if ("S_asymp" %in% index){
+     
+      S_asymp_sample = try(vegan::estimateR(mob_in$comm))
+      if (class(S_asymp_sample) == "try_error"){
+         warning("The Chao richness estimator cannot be calculated for all samples.")
+         S_asymp_sample <- rep(NA, nrow(mob_in$comm))
+      } else {
+         S_asymp_sample = S_asymp_sample["S.chao1",]
+         S_asymp_sample[!is.finite(S_asymp_sample)] <- NA
+      }
       
-      # random group level difference
-      group_bin_rand = sample(group_bin, replace = F)
-      abund_group_bin = aggregate(mob_in$comm, by = list(group_bin_rand), FUN = "sum")
-      group_bin_stats = apply(abund_group_bin[ ,-1], MARGIN = 1,
-                              FUN = calc_biodiv_single_group, n_rare = rarefy_groups)
-      colnames(group_bin_stats) <- abund_group_bin[,1]
-      delta_group_rand[i,] <- group_bin_stats[,"treatment"] - group_bin_stats[,"control"]
-   }
+      out$samples$S_asymp <- S_asymp_sample
+      
+      S_asymp_group <- try(vegan::estimateR(abund_group[,-1]))
+      if (class(S_asymp_group) == "try_error"){
+         warning("The Chao richness estimator cannot be calculated for all groups.")
+      } else {
+         S_asymp_group = S_asymp_group["S.chao1",]
+         S_asymp_group[!is.finite(S_asymp_group)] <- NA
+      }
+      
+      out$groups$S_asymp <- S_asymp_group
+  }
 
-   # p-value for difference among samples 
-   p_N       = sum(F_obs$N       <= F_rand$N) / nperm
-   p_S       = sum(F_obs$S       <= F_rand$S) / nperm
-   p_S_rare1 = sum(F_obs$S_rare1 <= F_rand$S_rare1) / nperm
-   p_S_rare2 = sum(F_obs$S_rare2 <= F_rand$S_rare2) / nperm
-   p_S_rare3 = sum(F_obs$S_rare3 <= F_rand$S_rare3) / nperm
-   p_S_asymp = sum(F_obs$S_asymp <= F_rand$S_asymp) / nperm
-   p_PIE     = sum(F_obs$PIE     <= F_rand$PIE) / nperm
-   p_ENS_PIE = sum(F_obs$ENS_PIE <= F_rand$ENS_PIE) / nperm
-   p_beta_S  = sum(F_obs$beta_S <= F_rand$beta_S) / nperm
-   p_beta_ENS_PIE  = sum(F_obs$beta_ENS_PIE <= F_rand$beta_ENS_PIE) / nperm
-   
-   pvalues_samples = data.frame(N       = p_N,
-                                S       = p_S,
-                                S_rare1 = p_S_rare1,
-                                S_rare2 = p_S_rare2,
-                                S_rare3 = p_S_rare3,
-                                S_asymp = p_S_asymp,
-                                PIE     = p_PIE,
-                                ENS_PIE = p_ENS_PIE,
-                                beta_S  = p_beta_S,
-                                beta_ENS_PIE  = p_beta_ENS_PIE
-                                )
+   # Probability of Interspecific Encounter (PIE)-------------------------------
+   if ("PIE" %in% index){ 
+      
+      plots_n01 = out$samples$N <= 1 # Hurlbert's PIE can only be calculated for two or more individuals
```

Comment:
It may be my misunderstanding - based on calc_PIE, it seems that PIE is 0 when out$sample$N = 0 but it is NA when out$sample$N = 1. Are both cases dropped?